ambry-quota/src/main/java/com/github/ambry/quota/capacityunit/AmbryCUQuotaSource.java

```diff
@@ -0,0 +1,215 @@
+  private final ScheduledExecutorService usageRefresher;
+  private final long aggregationWindowsInSecs;
+  private final AtomicBoolean isReady;
+  protected final AtomicReference<CapacityUnit> feUsage; // Ambry frontend's CU usage.
+
+  /**
+   * Constructor for {@link AmbryCUQuotaSource}.
+   * @param quotaConfig {@link QuotaConfig} object.
+   * @param accountService {@link AccountService} object.
+   * @throws IOException in case of any exception.
+   */
+  public AmbryCUQuotaSource(QuotaConfig quotaConfig, AccountService accountService) throws IOException {
+    feQuota = JsonCUQuotaDataProviderUtil.getFeCUCapacityFromJson(quotaConfig.frontendCUCapacityInJson);
+    cuQuota = new ConcurrentHashMap<>(
+        JsonCUQuotaDataProviderUtil.getCUQuotasFromJson(quotaConfig.resourceCUQuotaInJson, accountService));
+    cuUsage = new ConcurrentHashMap<>();
+    usageRefresher = Utils.newScheduler(1, REFRESHER_THREAD_NAME_PREFIX, true);
+    aggregationWindowsInSecs = quotaConfig.cuQuotaAggregationWindowInSecs;
+    isReady = new AtomicBoolean(false);
+    feUsage = new AtomicReference<>(null);
+  }
+
+  @Override
+  public synchronized void init() {
+    if (isReady.get()) {
+      LOGGER.warn("InMemoryCuQuotaSource is already initialized.");
+      return;
+    }
+    feUsage.set(new CapacityUnit());
+    cuQuota.keySet().forEach(key -> cuUsage.put(key, new CapacityUnit()));
+    usageRefresher.scheduleAtFixedRate(this::resetQuotaUsage, aggregationWindowsInSecs, aggregationWindowsInSecs,
+        TimeUnit.SECONDS);
+    isReady.set(true);
+  }
+
+  @Override
+  public boolean isReady() {
+    return isReady.get();
+  }
+
+  @Override
+  public Quota getQuota(QuotaResource quotaResource, QuotaName quotaName) throws QuotaException {
+    checkSupported(quotaName, quotaResource);
+
+    final String resourceId = quotaResource.getResourceId();
+    assertResourceId(resourceId);
+    return new Quota<>(quotaName, cuQuota.get(resourceId).getQuotaValue(quotaName), quotaResource);
+  }
+
+  @Override
+  public float getUsage(QuotaResource quotaResource, QuotaName quotaName) throws QuotaException {
+    checkSupported(quotaName, quotaResource);
+    final String resourceId = quotaResource.getResourceId();
+    assertResourceId(resourceId);
+    double usage = 0;
+    if (cuUsage.containsKey(resourceId)) {
+      usage = cuUsage.get(resourceId).getQuotaValue(quotaName);
+    }
+    return QuotaUtils.getUsagePercentage(cuQuota.get(resourceId).getQuotaValue(quotaName), usage);
+  }
+
+  @Override
+  public void chargeUsage(QuotaResource quotaResource, QuotaName quotaName, double usageCost) throws QuotaException {
+    checkSupported(quotaName, quotaResource);
+    final String resourceId = quotaResource.getResourceId();
+    assertResourceId(resourceId);
+    if (quotaName == QuotaName.READ_CAPACITY_UNIT) {
+      cuUsage.get(resourceId).incrementRcu((long) Math.ceil(usageCost));
```

Comment:
There is a case where assertResourceId could return successfully but you get an NPE from cuUsage if updateNewQuotaResources is still running. Maybe you can protect against this with a similar existence check to getUsage.

Reply:
fixed.